src/core/net/tcp6.cpp

```diff
@@ -166,15 +296,98 @@ Error Tcp::Listener::StopListening(void)
 
 Error Tcp::Listener::Deinitialize(void)
 {
-    return kErrorNotImplemented;
+    Error     error;
+    Listener *prev;
+
+    Tcp &tcp = GetInstance().Get<Tcp>();
+    SuccessOrExit(error = tcp.mListeners.Find(*this, prev));
+    tcp.mListeners.PopAfter(prev);
+    SetNext(nullptr);
+
+exit:
+    return error;
 }
 
-Error Tcp::ProcessReceivedSegment(Message &aMessage, MessageInfo &aMessageInfo)
+Error Tcp::HandleMessage(ot::Ip6::Header &aIp6Header, Message &aMessage, MessageInfo &aMessageInfo)
 {
+    OT_UNUSED_VARIABLE(aIp6Header);
     OT_UNUSED_VARIABLE(aMessage);
     OT_UNUSED_VARIABLE(aMessageInfo);
 
-    return kErrorNotImplemented;
+    Error error = kErrorNotImplemented;
+
+    for (Endpoint *active = mEndpoints.GetHead(); active != nullptr; active = active->GetNext())
+    {
+    }
+
+    for (Listener *passive = mListeners.GetHead(); passive != nullptr; passive = passive->GetNext())
+    {
+    }
+
+    return error;
+}
+
+void Tcp::HandleTimer(Timer &aTimer)
+{
+    OT_ASSERT(&aTimer == &aTimer.GetInstance().Get<Tcp>().mTimer);
+    otLogDebgTcp("Main TCP timer expired");
+    aTimer.GetInstance().Get<Tcp>().ResetTimer(true);
+}
+
+void Tcp::ResetTimer(bool aFirePendingTimers)
+{
+    TimeMilli now                        = TimerMilli::GetNow();
+    bool      pendingTimer               = false;
+    TimeMilli earliestPendingTimerExpiry = now.GetDistantFuture();
+
+    /*
+     * The timer callbacks could potentially set/reset/cancel timers, so we
+     * could wind up re-entering this function in a timer callback. If that
+     * happens, the second call can be skipped, since the original call will
+     * anyway reset mTimer after firing the callbacks. The mFiringTimers
+     * variable is how this is enforced.
+     */
+    if (aFirePendingTimers)
```

Comment:
I think that the current version of the code is fine. The timers callbacks don't ever call `ResetTimer()` anymore, so don't have to worry about `mTimer` being mutated by the callbacks. They may change other timer values on the same TCP endpoint, but a timer callback in one TCP endpoint won't ever affect timers on another TCP endpoint. So we can just iterate over the linked list once. For each endpoint, we iterate over its timers twice: once to fire any callbacks, and a second time to determine the earliest timer.

Reply:
I think better to use `FireAtIfEarlier()` at the end. I see you are using `FiretAt()` at end of `FireAllPendingTimers()` method.

When handling expired timers, we can invoke callbacks and I think potentially even go out of OT code (callback to user and app layer code) which can then call any other APIs. When calling a callback we should assume things may change after we return from it.

Another related point: when iterating over end points `for (Endpoint *endpoint = mEndpoints.GetHead(); endpoint != nullptr; endpoint = endpoint->GetNext())` we are assuming `endpoint` remains valid and on list (so we can use `endpoint->GetNext()` to get to the next on the list. Can it be possibly removed from one of the timer callbacks?

If so, we need to handle this in the code (one way to do this when iterating over linked-list is to save `next = endpoint->GetNext()` at the top of the loop in a variable and use `endpoint = next` in loop iteration step, e.g see `MeshForwarder::UpdateReassemblyList()`). But there may be other situations to consider after calling `endpoint->FirePendingTimers(now);` (if it can change `endpoint`?).

Reply:
The TCPlp timer callbacks are TCP functions. Most user-specified callbacks are called in response to receiving a packet, which is why I hadn't thought about this earlier... but thinking about what you said, there is one callback--- `connection_lost`---that might be called in response to a timer event, if it causes the whole TCP connection to time out.

So I agree that we need to write this more defensively, since the callback might enter user code, which may update timers on the same TCP endpoint and/or the other endpoints, or even mutate the linked list. This is what I think we can assume:
* Any newly set timers won't have expired yet, since TCPlp will only set timers to expire in the future (including some minimum delay)
* Any newly created TCP endpoints won't have any expired timers (corollary of above), but they might have active timers (if the user code ended up calling `otTcpEndpointConnect` or something).

I think that either of the two approaches you suggested (the first approach is to `FireAtIfEarlier`, the second approach is to iterate over the `Endpoint`s twice) would work. For now, I've just used `FireAtIfEarlier` since that's more efficient.